### ISSUE TITLE
Support dependency injection into instances

### ISSUE THREAD

**Issue by cressie176** [sequelize/sequelize#11767]
## Issue Description
I would like to the ability to inject dependencies into model instances at the point where Sequelize creates them.

### Is your feature request related to a problem? Please describe.
I often find myself wanting to use dependencies (e.g. api or messaging clients) from my model instances, e.g.
```js
static async function ensureUser(username) {
  let user = await User.findOneByCriteria({ where: { username }});
  if (user) return user;

  user = await this.userApi.fetchUserByEmail(username);
  return User.save(user);
}
// n.b. There's a race condition in the above, this is just a simple example to demonstrate the dependency issue*
```    

I haven't found a nice way to achieve with Sequelize because there doesn't appear to be a way to inject dependencies into model instances. 

### Describe the solution you'd like
I was wondering whether a `afterConstruct` hook could be used to solve this. I understand it might be slow for large record sets, but hopefully not that slow if all it was doing was fetching a pre-created dependency.

```js
function initModels(dependencies) {
  User.init({
    username: DataTypes.STRING,
  }, {
    hooks: {
      afterConstruct: (user, options) => {
        user.userApi = dependencies.userApi;
        user.messageBroker = dependencies.messageBroker
        // n.b. assignment might need to support async factories, although caution should be taken as this could be extremely slow for large record sets
      },
    },
    sequelize
    // An alternative solution might be to support a dependencies object here
  });
}
```

### Why should this be in Sequelize
One of the primary use cases for Sequelize (or any ORM) is to support rich domain models, which collocate behaviour and state. Not being able to inject dependencies undermines this.

### Describe alternatives/workarounds you've considered
The options I've considered are:
* Requiring the dependency / using singletons. This is infeasible if the dependency has async setup or requires its own dependencies. Also makes TDD harder.
* Passing the dependencies as arguments to the function, e.g. `ensureUser(username, userApi)`. This adversely affects readability (by increasing the number of function arguments) and potentially means passing dependencies through several levels of function call, before they're actually used 
* Copying the constructor approach you use for injecting `this.sequelize` (this is what we're currently using as a workaround)

### Is this issue dialect-specific?

- [x] No. This issue is relevant to Sequelize as a whole.

### Would you be willing to resolve this issue by submitting a Pull Request?

- [x] Yes, I have the time and I know how to start (but may also benefit from guidance).

**Comment by cressie176** [comment]
After reading another issue it appears that hooks are not triggered for "included" entities. This would be a problem for my proposed solution. The constructor/this.sequelize approach may a more feasible option.

**Comment by papb** [comment]
Thanks for the request, I remember that once I wanted an `afterConstruct` hook too...

**Comment by papb** [comment]
Would you like to send a PR?

**Comment by cressie176** [comment]
Hi @papb, we'd be happy to work on a PR. Any thoughts about hooks not being fired for nested / included entities?

**Comment by papb** [comment]
> we'd be happy to work on a PR.

Nice!! Thanks!

> Any thoughts about hooks not being fired for nested / included entities?

I think that making all hooks fire for nested entitites is too much work... I prefer the idea of an `afterConstruct` hook, and of course making this hook fire for nested entities (without having to make all existing hooks also fire for nested entities)... In other words, just add an `afterConstruct` hook wherever Sequelize calls `new` for a model, I guess?

**Comment by cressie176** [comment]
Thanks @papb. We have a hackday coming up soon. We'll work on the PR then.

**Comment by cressie176** [comment]
Hi @papb,

We've started to dig into this and have hit a bit of a snag. The most obvious place to run this hook would be from [Model.build](https://github.com/sequelize/sequelize/blob/master/lib/model.js#L2169), since this function is called from anywhere that constructs instances. We tested this and it works well, however this approach would be inconsistent with the existing pattern for bulk operations.

By default, bulk operations do not run individual hooks, but when enabled, the individual hooks are [run separately in a loop](https://github.com/sequelize/sequelize/blob/master/lib/model.js#L2620-L2632), **after** the instances have been created. This will be difficult to replicate with the proposed approach. What are your thoughts on always running the hook from `Model.build` regardless of whether it is called in a bulk or individual context?
